### ISSUE TITLE
[Strix 10/20] Route HIP prefill to Vulkan decode without replaying Qwen3.8 state

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12540]
# Route HIP prefill to Vulkan decode without replaying Qwen3.8 state

Parent campaign: #11572. Frozen comparison ladder: #12093. Matched physical comparison gate: #12098.

## Goal

Hand one completed Qwen3.8 prefill state allocation from HIP ownership to Vulkan ownership without replay.

## Context

- **For:** Strix Halo users who need both long-prompt throughput and low single-stream decode latency.
- **Problem:** the strongest AMD backend can differ by phase, but fak cannot hand off an already-prefilled Qwen3.8 session between HIP and Vulkan without replay or host reconstruction.
- **Today:** backend selection is session-wide; repository queries find native profile and benchmark records but no phase router or cross-API KV/GDN state handoff. Existing prefill and decode work therefore compete as exclusive whole-session choices.
- **Better because:** the APU's unified physical memory can support an explicit ownership transition while fak retains state identity, correctness, and accounting.
- **Witness:** a deterministic synthetic handoff proves no replay and unchanged state digests; a real pp2048+decode run compares fixed-Vulkan with HIP-prefill/Vulkan-decode.

Centrality: **Core**, priority tier **2 — fak serving only**. P1-P4: combines two phase-specific levers; owns a concrete state boundary; fails closed instead of replaying; is reversible to single-backend execution.

## Comparison context — not a victory claim

The frozen study records Halogen custom-HIP Qwen3.8 prefill at **620/710/566 t/s** for pp512/2048/32768 and DFlash2 decode at **31.71 t/s mean**; nabe llama.cpp Vulkan+MTP at **16.65 t/s c1** through **30.61 t/s c16**. Local sealed fak-native runs are **0.34/0.37 t/s** and `NOT_COMPARABLE`. A pinned HowQuant/nabe HIP-vs-Vulkan receipt (`0ee4`) and an older phase-sensitive Qwen GDN receipt show backend leadership can reverse between prefill and decode; route: **INSPIRE-ONLY**, because neither proves a safe fak state handoff.

## Prior art and self-query

- Study receipts: `study_fea6c1bc31f94fcee01ec566289e7ca509c28e7a982068dbde17b36e33b333e3` and `study_a8e74aeef35f4f79a52d2af0a1d1f6ef8c4830eabdf419db4ddf3a2052c08c1a`.
- Self-query: `fak capabilities "AMD phase router HIP prefill Vulkan decode state handoff" --json` plus `rg -n "phase.*route|HIP.*Vulkan|Vulkan.*HIP" internal/model internal/compute` => **ABSENT**: no executable router or ownership-transfer contract.
- External results choose the hypothesis only. The implementation must be derived from fak's state layout and backend ownership contracts, not copied from a competitor.

## Dependencies and pickup rule

- Start blocked by: #12539
- Coordinates with: #12395
- Coordinates with: #12397
- Promotion requires: #12098

The rank-09 HIP panel and its registered allocation owner must exist before pickup; an injected fixture alone cannot establish the real cross-API ownership seam. #12395/#12397 own adjacent measurement/scheduling work, not state ownership.

## One bounded deliverable

Implement one explicit HIP-to-Vulkan state-allocation ownership transfer in at most these three related files:

- `internal/compute/amd_phase_handoff.go`
- `internal/compute/vulkan_hip_interop.cpp`
- `internal/compute/amd_phase_handoff_test.go`

The leaf must export/import one admitted attention-KV plus GDN-state allocation through explicit external-memory handles, transfer ownership through an explicit external semaphore, preserve token position and state digest, and record every handle/import/synchronization event. It must reject missing compatibility metadata or unsupported driver capabilities. No prompt replay, CPU reconstruction, llama.cpp route, implicit fallback, or per-token backend oscillation is permitted.

## Deterministic software witness

```text
go test -tags "rocm vulkan" -count=1 -run TestAMDPhaseHandoffPreservesQwen38StateWithoutReplay ./internal/compute
```

The test must assert identical pre/post state digests and next-token logits against fixed-backend execution, zero replayed tokens, one ownership transition, complete transfer accounting, and a typed refusal on incompatible state geometry.

## Exact Strix hardware witness and promotion rule

```text
fak validate --strix --subkernels=qwen38-phase-route --ablate=vulkan-only,hip-prefill-vulkan-decode
```

Use the same pp2048 prompt and fixed generated-token boundary, one warm-up, and **N>=5 measured repetitions** per arm with pinned artifact, quantization, clocks/power, drivers, and revision. Report prefill, handoff, decode, and total medians plus every run. Promotion requires greedy output-token and selected-logprob parity, state-digest equality at handoff, `engine=fak-native`, `fallback_count=0`, and **>=5% lower median total time** for the routed arm; otherwise retain Vulkan-only and record the rejection. Comparison-point claims require #12098.

**Comment by anthony-chaudhary** [comment]
Coordination from #12539 source owner: #12539 does not provide the hard prerequisite described here. Its current candidate is uncommitted and component-only; rocmQwen38Backend.NewKV explicitly refuses, and the Into ABI owns resident Q/K/V/output tensors only. It creates no completed model prefill session, attention-KV plus GDN state allocation owner, exportable external-memory handle, or external-semaphore lifetime contract. The separate existing rocmBackend has an F32 KV store, but inspection of its backend/KV/header/HIP sources found no export/import or GDN handoff ownership seam satisfying this issue. A component output buffer must not be substituted for the required state allocation. #12540 implementation remains queued pending an actual model-state owner/export-lifetime contract; no overlapping edits or hardware action were taken.

**Comment by anthony-chaudhary** [comment]
Bounded contract review at public trunk 430fd2b2c71fd23c6600c41f0f4b32eaceb56401 (2026-09-12): implementation remains QUEUED under the existing Start blocked by #12539 rule. No code paths, device window, or peer lease were taken; no commit or test/physical PASS is claimed.

Verified prerequisite gap:
- #12539 is OPEN with exactly three uncommitted component candidate files. Its revised scope explicitly excludes full-model logits/KV/greedy continuation and HIP-to-Vulkan handoff. Its final source-only receipt is https://github.com/anthony-chaudhary/fak/issues/12539#issuecomment-5646931675 . Component qualification alone therefore cannot fulfill this issue's registered attention-KV plus GDN allocation-owner prerequisite.
- #12598's witness owner confirms zero physical A/B runs and no KEEP/REJECT receipt. That witness covers resident Q/K/V/output component tensors only, not model-state ownership, external handles/semaphores, or continuation. Unsupported D2D observability must remain unsupported, never measured zero.
- At the reviewed trunk, rocm_kv.go owns separate per-layer K/Kraw/V allocations; rocm_kernels.hip allocates/frees with hipMalloc/hipFree. rocm_backend.go buffer lifetime fields do not supply an export/import ownership transaction. The three #12540 target files are absent, and source queries found no HIP/Vulkan external-memory/semaphore bridge.
- The related companion ticket fak-private#676 records software admission/receipt validation, with physical ownership and comparative gates still unchecked. It does not supply the public allocation owner.
- The qwen38-phase-route selector and TestAMDPhaseHandoffPreservesQwen38StateWithoutReplay are absent at this trunk. The issue's proposed hardware invocation is a future acceptance contract, not an executable qualification receipt. The eventual dual-profile integration must also respect #12539's additional rocm_qwen38_prefill tag. Serialize validator/shim/registry integration with its current owners.

Resume contract:
1. Name and land a real producer owner for the completed attention-KV plus GDN state, with the admitted layout/dtype/offsets, token position, artifact/device identity, lifetime, and external-handle capability. Reconcile which issue owns this prerequisite: the current component-only #12539 contract does not include it. Keep this issue start-blocked until that is explicit and witnessed; do not substitute fixtures.
2. Only then acquire the atomic issue/path leases for amd_phase_handoff.go, vulkan_hip_interop.cpp, and amd_phase_handoff_test.go. Define one transfer transaction with producer completion signaling, consumer wait/import, exactly-once handle/import/semaphore accounting and cleanup, and typed incompatibility refusal before ownership changes. Failed transfer must not leave two writers or free storage still in use.
3. Prove state digest/token-position preservation, next-token logits and greedy continuation parity, zero replay, exactly one ownership transition, and complete transfer accounting through the real owner seam.
4. Keep promotion open until a source-bound physical paired pp2048+decode run supplies warm-up plus N>=5 measured repetitions per arm, every prefill/handoff/decode/total timing and median, engine=fak-native, fallback_count=0, selected-logprob/output parity, and at least 5% lower median total time. Otherwise retain Vulkan-only; #12098 continues to gate comparison claims.

Next checkable step: reconcile and witness the public full-state allocation-owner prerequisite, then rerun pickup admission. This review adds no HIP performance or handoff delivery credit.
